### DISCUSSION TITLE
Day 1: Fundamentals of Software Architecture - Session 4: Influencing Factors in Architecture Design

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/iSAQB-CPSA-F#5] [Notes]
**Day 1: Fundamentals of Software Architecture**

**Session 4: Influencing Factors in Architecture Design**

**Duration:** 60 minutes

**Objective:** Equip participants with a comprehensive understanding of the various factors that influence software architecture design, emphasizing their significance in the automotive industry.

**Session Content:**

1. **Introduction to Influencing Factors in Architecture Design:**

   - **Definition:**
     - Influencing factors are the diverse elements that impact the decisions made during the design of a software architecture.

   - **Importance:**
     - Recognizing and analyzing these factors is crucial for creating architectures that are robust, scalable, and aligned with business objectives.

2. **Categories of Influencing Factors:**

   - **Technical Factors:**
     - **System Requirements:** Functional and non-functional requirements that the system must fulfill.
     - **Technology Stack:** Selection of programming languages, frameworks, and tools.
     - **Integration Needs:** Necessity to interface with existing systems or third-party services.

   - **Business Factors:**
     - **Business Goals:** Objectives such as market expansion, cost reduction, or innovation.
     - **Budget Constraints:** Financial limitations impacting technology choices and resource allocation.
     - **Time-to-Market:** Urgency in delivering the product to gain competitive advantage.

   - **Organizational Factors:**
     - **Team Expertise:** Skill sets and experience levels of the development team.
     - **Development Processes:** Adopted methodologies like Agile, Scrum, or Waterfall.
     - **Organizational Structure:** Hierarchy and communication channels within the organization.

   - **Environmental Factors:**
     - **Regulatory Compliance:** Adherence to industry standards and legal requirements.
     - **Market Trends:** Current trends and customer expectations influencing design choices.
     - **Competitive Landscape:** Strategies employed by competitors that may affect architectural decisions.

3. **Detailed Exploration of Influencing Factors:**

   - **Technical Factors:**
     - **System Requirements:**
       - *Functional Requirements:* Specific behaviors or functions the system must perform.
       - *Non-Functional Requirements:* Attributes such as performance, scalability, security, and usability.
     - **Technology Stack:**
       - *Programming Languages:* Choice between languages like Java, C++, or Python based on project needs.
       - *Frameworks and Libraries:* Utilization of existing tools to accelerate development.
     - **Integration Needs:**
       - *Legacy Systems:* Challenges in interfacing with outdated systems.
       - *Third-Party Services:* Considerations for APIs and external service dependencies.

   - **Business Factors:**
     - **Business Goals:**
       - *Market Expansion:* Designing for scalability to accommodate growth.
       - *Cost Reduction:* Implementing efficient architectures to minimize operational expenses.
     - **Budget Constraints:**
       - *Resource Allocation:* Balancing cost with quality and performance.
       - *Technology Investments:* Decisions on purchasing licenses or developing in-house solutions.
     - **Time-to-Market:**
       - *Development Speed:* Choosing architectures that enable rapid development cycles.
       - *Competitive Advantage:* Delivering features ahead of competitors to capture market share.

   - **Organizational Factors:**
     - **Team Expertise:**
       - *Skill Sets:* Aligning architecture with the team's strengths and experience.
       - *Training Needs:* Identifying areas where additional training may be required.
     - **Development Processes:**
       - *Methodologies:* Impact of Agile vs. Waterfall on architectural flexibility.
       - *Tooling:* Selection of tools that support the chosen development process.
     - **Organizational Structure:**
       - *Communication Channels:* Ensuring effective collaboration among stakeholders.
       - *Decision-Making Hierarchy:* Understanding who has authority over architectural decisions.

   - **Environmental Factors:**
     - **Regulatory Compliance:**
       - *Industry Standards:* Adhering to protocols like ISO 26262 in automotive software.
       - *Legal Requirements:* Ensuring data privacy and security regulations are met.
     - **Market Trends:**
       - *Customer Expectations:* Designing features that align with current user demands.
       - *Technological Advancements:* Incorporating emerging technologies to stay competitive.
     - **Competitive Landscape:**
       - *Benchmarking:* Analyzing competitors' offerings to inform architectural choices.
       - *Differentiation Strategies:* Designing unique features to stand out in the market.

4. **Case Study: Volkswagen's Digital Transformation**

   - **Background:**
     - Volkswagen embarked on a digital transformation journey to become a leading software-driven automotive company.

   - **Influencing Factors:**
     - **Technical Factors:**
       - *System Requirements:* Need for a unified software platform across all vehicle models.
       - *Technology Stack:* Adoption of cloud technologies and development of the Volkswagen Operating System (VW.OS).
     - **Business Factors:**
       - *Business Goals:* Enhance customer experience through connected services.
       - *Budget Constraints:* Significant investment in software development and talent acquisition.
     - **Organizational Factors:**
       - *Team Expertise:* Building in-house software development capabilities.
       - *Development Processes:* Transitioning to Agile methodologies to increase flexibility.
     - **Environmental Factors:**
       - *Regulatory Compliance:* Ensuring compliance with automotive safety and data privacy regulations.
       - *Market Trends:* Responding to the growing demand for electric and autonomous vehicles.

   - **Outcome:**
     - Development of the Volkswagen Industrial Cloud in collaboration with AWS to connect 124 factory sites to a single architecture, bringing efficiencies to manufacturing and logistics processes. citeturn0search15

5. **Exercises and Activities:**

1. **Guided Exercise: Identifying Influencing Factors**

   - **Objective:** Enhance participants' ability to recognize and analyze various factors influencing architectural design decisions.

   - **Instructions:**

     - Present participants with a scenario involving the development of a new automotive software system.

     - Ask them to identify potential influencing factors across technical, business, organizational, and environmental categories.

     - Encourage discussion on how each factor might impact architectural decisions.

   - **Example Scenario:**

     - **Context:** Developing an advanced driver-assistance system (ADAS) for a new vehicle model.

     - **Expected Responses:**

       - **Technical Factors:**

         - *System Requirements:* Real-time processing of sensor data.

         - *Technology Stack:* Integration of machine learning algorithms.

       - **Business Factors:**

         - *Business Goals:* Enhancing vehicle safety to meet market demands.

         - *Budget Constraints:* Allocating funds for research and development.

       - **Organizational Factors:**

         - *Team Expertise:* Availability of engineers skilled in AI and sensor technologies.

         - *Development Processes:* Adoption of Agile methodologies for iterative development.

       - **Environmental Factors:**

         - *Regulatory Compliance:* Adherence to safety standards and regulations.

         - *Market Trends:* Growing consumer interest in autonomous driving features.

2. **Unguided Exercise: Prioritizing Influencing Factors**

   - **Objective:** Develop participants' skills in assessing and prioritizing influencing factors based on their impact on architectural decisions.

   - **Instructions:**

     - Provide participants with a list of influencing factors related to a specific project.

     - Ask them to rank these factors in order of importance and justify their prioritization.

   - **Example List:**

     - *Time-to-Market*

     - *Integration with Legacy Systems*

     - *Team Expertise*

     - *Regulatory Compliance*

     - *Budget Constraints*

   - **Expected Outcome:**

     - Participants will discuss and determine the most critical factors influencing the project's architecture, fostering critical thinking and decision-making skills.

3. **Case Study Analysis: Volkswagen's Digital Cockpit Development**

   - **Objective:** Analyze a real-world example to understand how various influencing factors shape architectural decisions.

   - **Instructions:**

     - Present the case study of Volkswagen's development of a digital cockpit for their vehicles.

     - Ask participants to identify the influencing factors involved and discuss how they impacted the architectural design.

   - **Discussion Points:**

     - **Technical Factors:**

       - *System Requirements:* High-resolution displays and responsive user interfaces.

       - *Technology Stack:* Integration of touchscreens, voice recognition, and connectivity features.

     - **Business Factors:**

       - *Business Goals:* Enhancing user experience to differentiate from competitors.

       - *Time-to-Market:* Rapid development to align with vehicle launch schedules.

     - **Organizational Factors:**

       - *Team Expertise:* Collaboration between software developers and hardware engineers.

       - *Development Processes:* Implementation of Agile practices to accommodate iterative design.

     - **Environmental Factors:**

       - *Regulatory Compliance:* Ensuring compliance with safety and usability standards.

       - *Market Trends:* Responding to consumer demand for advanced in-car technology.

**Sample Exam Questions:**

1. **Multiple Choice:**

   - **Question:** Which of the following is a technical factor influencing software architecture design?

     - A) Business Goals

     - B) Team Expertise

     - C) Regulatory Compliance

     - D) System Requirements

   - **Answer:** D) System Requirements

2. **True or False:**

   - **Statement:** Budget constraints are considered a technical factor in architectural design.

   - **Answer:** False

3. **Short Answer:**

   - **Question:** Explain how regulatory compliance can influence software architecture decisions in the automotive industry.

   - **Answer:** Regulatory compliance requires adherence to industry standards and legal requirements, such as safety protocols and data privacy laws. This necessitates incorporating specific features and constraints into the software architecture to ensure the system meets all mandated guidelines, impacting design choices and implementation strategies.

**Conclusion:**

Understanding the diverse factors that influence software architecture design is essential for developing systems that are robust, compliant, and aligned with business objectives.

**Comment by shruti-kirti-vwits** [comment]
Volkswagen's Digital Cockpit Development

Factors:

Technical:
1. Good-sized cache, high processing chips, comprehensive design for the hardware assembly should be identified.
2. Latency should be low, hence use most responsive & on the fly processing technology stack. 

Business:
1. The dashboards should be unified across all models.
2. The dashboard should have room for scalability for adding new features

Org:
1. Development process should be Agile 
2. Follow lean channel for the organizations structure.
3. Team should be competent with the tech stack finalized

Environment:
1. Prefer LCD or OLED for better vision.
2. Check on the trends like Pedestrian Detection | Vehicle proximity, Augmented systems

**Comment by Sath0111** [comment]
Digital Cockpit : Influencing Factors :
	
1. Technical Factors : The design must be simple to make it easy to visualize or use during car driving, must be able to communicate to the internal data to fetch and display. The technology must have the possibility to have enhanced UI, communicate with different internal systems and must also be responsive. If 1 technology is not able to fulfill all these, it might result in complex Integrations.
	
2. Business Factors : Who is the target population using this Car largely, depending on them the UI can be more advanced or simple; Depending on the budget and cost the design will directly be impacted. 
	
3. Org Factors : What resources and skills do we have to build this; Increase in cost due to trainings that would be required for resources.

4. Environmental Factors : Does it adhere to different regulatory constraints of the countries. Making the experience unique for users would result in more investment costs.

**Comment by VishwasRS** [comment]
Project: Group-wide Central Procurement Solution on SAP 

Business Factor: 
Budget constraint due to Cost cutting

Technical Factor : 
Technology stack : SAP Central Procurement HUB solution on HANA
Technical integration : Interface issues with Legacy system


Organizational Factors: 
Large group with decentralized systems across worldwide 70+ companies 
Customized business process for each country & company.

Environmental Factors:
Regulatory Compliance: Compliance with safety and data privacy regulations & SAP Audits.

**Comment by Sps8786** [comment]
Technical Factor :
Technology stack : No complex integration. Easily upgraded to newer version. User friendly UI design. 
Organizational Factors: 
Collaboration is the process of working with other people on a project. The benefits of collaboration are that you are exposed to different ideas and perspectives about the project, your work and the way the design could proceed.
Environmental Factors: 
Sustainability, Is the product good for people, the planet and the environment.
Business Goals:
Project cost and fees, The overall cost of a project and the budget you need to work within, and fees for your services.
Scope of services (design) , What exactly you are expected to do and provide in terms of design, documentation and deliverables.

**Comment by Mangesh1206** [comment]
Factors Influencing : VW Digital cockpit

- Business factor: 
1. Budget constraint: Production cost and its impact on per unit selling price.

- Technical factor: 

1. Technology stack: integration AI features, Voice control features
2. System/hardware: dimensions of the Cluster

- Enviromental factors
1. Compliance: - Adhering to country specific Government Regulations.
2. Market trends: Comparision with other brands in same segment the features they offer.

- Org factor:
1. Technical capability- Decision to make or purchase the software.
